packages/vite/src/node/fsUtils.ts

```diff
@@ -0,0 +1,416 @@
+  onDirectoryUnlink?: (file: string) => void
+}
+
+// An implementation of fsUtils without caching
+export const commonFsUtils: FsUtils = {
+  existsSync: fs.existsSync,
+  isDirectory,
+
+  tryResolveRealFile,
+  tryResolveRealFileWithExtensions,
+  tryResolveRealFileOrType,
+}
+
+const cachedFsUtilsMap = new WeakMap<ResolvedConfig, FsUtils>()
+export function getFsUtils(config: ResolvedConfig): FsUtils {
+  let fsUtils = cachedFsUtilsMap.get(config)
+  if (!fsUtils) {
+    if (config.command !== 'serve' || !config.server.fs.cachedChecks) {
+      // cached fsUtils is only used in the dev server for now, and only when the watcher isn't configured
+      // we can support custom ignored patterns later
+      fsUtils = commonFsUtils
+    } /* TODO: Enabling for testing, we need to review if this guard is needed
+    else if (config.server.watch === null || config.server.watch?.ignored) {
+      config.logger.warn(
+        colors.yellow(
+          `${colors.bold(
+            `(!)`,
+          )} server.fs.cachedChecks isn't supported if server.watch is null or a custom server.watch.ignored is configured\n`,
+        ),
+      )
+      fsUtils = commonFsUtils
+    } */ else {
+      fsUtils = createCachedFsUtils(config)
+    }
+    cachedFsUtilsMap.set(config, fsUtils)
+  }
+  return fsUtils
+}
+
+type DirentsMap = Map<string, DirentCache>
+
+type DirentCacheType =
+  | 'directory'
+  | 'file'
+  | 'symlink'
+  | 'error'
+  | 'directory_maybe_symlink'
+  | 'file_maybe_symlink'
+
+interface DirentCache {
+  dirents?: DirentsMap
+  type: DirentCacheType
+}
+
+function readDirCacheSync(file: string): undefined | DirentsMap {
+  let dirents: fs.Dirent[]
+  try {
+    dirents = fs.readdirSync(file, { withFileTypes: true })
+  } catch {
+    return
+  }
+  return direntsToDirentMap(dirents)
+}
+
+function direntsToDirentMap(fsDirents: fs.Dirent[]): DirentsMap {
+  const dirents = new Map<string, DirentCache>()
+  for (const dirent of fsDirents) {
+    // We ignore non directory, file, and symlink entries
+    const type = dirent.isDirectory()
+      ? 'directory'
+      : dirent.isSymbolicLink()
+        ? 'symlink'
+        : dirent.isFile()
+          ? 'file'
+          : undefined
+    if (type) {
+      dirents.set(dirent.name, { type })
+    }
+  }
+  return dirents
+}
+
+function ensureFileMaybeSymlinkIsResolved(
+  direntCache: DirentCache,
+  filePath: string,
+) {
+  if (direntCache.type !== 'file_maybe_symlink') return
+
+  const isSymlink = fs
+    .lstatSync(filePath, { throwIfNoEntry: false })
+    ?.isSymbolicLink()
+  direntCache.type =
+    isSymlink === undefined ? 'error' : isSymlink ? 'symlink' : 'file'
+}
+
+export function createCachedFsUtils(config: ResolvedConfig): FsUtils {
+  const { root } = config
+  const rootDirPath = `${root}/`
+  const rootCache = { type: 'directory' as DirentCacheType } // dirents will be computed lazily
+
+  const getDirentCacheSync = (parts: string[]): DirentCache | undefined => {
```

Comment:
I tried using a flat map of directory -> dirents/cache, but I ended up going with the current design to eliminate realpath calls (even when `preserverSymlinks` is false). The idea is that if we use `readdirSync` with `withFileTypes: true` we get for each dirent if it is a directory, file, or symlink (could be either). We don't expand symlinks.

When we are checking for a file path, we start from the root and iterate to the parent dir:
- if any directory in its path is a symlink, we fallback to the uncached version that ends up doing a realpath.
- if there aren't symlinks at any level, it means we can return the path without doing a realpath call.
This is the reason we gain so much with this PR, but to make this work, we need to expand the dirents at all levels. I was worried too at first, but given that we end up reusing most of the levels anyways, only the first calls are a bit slower. `readdirSync` as used right now seems very fast in the profiles I saw so far.

Reply:
I don't quite understand why a flat map would have an issue with realpath? Wouldn't the checks become:

```js
const dirents =  cacheMap.get(path.dirname(p)) ?? []
const isExist = dirents.some(d => d.name === path.basename(p))
```

We could keep the symlink checks in place, but only swap out how we cache things so it's more lazy.

Reply:
I just tried it again to check in case I interpreted `readdirSync` withFileTypes in the wrong way.
If you have a symlinked directory like:
```  
root/
  symlinked_dir/
    nested_in_symlink_regular_dir (...)
    nested_in_symlink_regular_file
  regular_dir/
    nested_regular_dir (...)
    nested_regular_file
  regular_file
```
Calling `readdirSync('./', { withFileTypes: true })` will give you:
```
  { name: 'symlinked_dir', 
    isDirectory() -> false, isFile() -> false, isSymbolicLink() -> true },
  { name: 'regular_dir',
    isDirectory() -> true, isFile() -> false, isSymbolicLink() -> false },
  { name: 'regular_file',
    isDirectory() -> false, isFile() -> true, isSymbolicLink() -> false },
```
Calling `readdirSync('./symlinked_dir', { withFileTypes: true })` will give you:
```
  { name: 'nested_in_symlink_regular_dir', 
    isDirectory() -> true, isFile() -> false, isSymbolicLink() -> false },
  { name: 'nested_in_symlink_regular_file',
    isDirectory() -> true, isFile() -> false, isSymbolicLink() -> false },
```
So if we need to know if a file needs the realpath call, we need to call `readdirSync` on every level. Calling it only on the last level doesn't provide that information. So we can't do it lazier than we have now if we want to avoid the realpath calls AFAICS.
The cache could still be flat, but given that we have to iterate each level to do the symlink check I think it wouldn't help codewise.

Reply:
By the way, we are only expanding from `root`. One thing that is missing that we should implement later is doing a `root === realpathSync(root)` guard to check if the root itself has a symlinked parent dir. In that case, we should fallback to the common fs utils if `preserveSymlinks` is false because realpath will be needed all the time. Or if having a symlinked root is a common enough case, we could have a cached fs tree version that keeps the checks optimizations (but still does the realpath call at the end).

Reply:
Added the guard for symlinked root at https://github.com/vitejs/vite/pull/15279/commits/7c562d236008a851c067f679da94497189752879

Reply:
Thanks for the explanation. That makes a lot of sense to me now 👍 I didn't know that symlink with directories would work that way. I'm fine with a graph based approach for now then and we can continue experimenting it.